### PR TITLE
Cleaner implementation of the equivalency validator

### PR DESCRIPTION
**Pull request by dennisdoomen**
While preparing a talk about maintainable code, I realized that the `EquivalencyValidator` was the perfect candidate for refactoring.

### CODE REVIEW

File: Src/FluentAssertions/Equivalency/EquivalencyValidator.cs [hunk 1]
```diff
@@ -44,87 +43,89 @@ public void AssertEquality(EquivalencyValidationContext context)
...
-            if (!isComplexTypeMap.TryGetValue(type, out bool isComplexType))
+            if (!shouldRecurse)
             {
-                isComplexType = !type.OverridesEquals();
-                isComplexTypeMap[type] = isComplexType;
+                AssertionScope.Current.FailWith("The maximum recursion depth was reached.  ");
             }
 
-            return isComplexType;
+            return shouldRecurse;
```
Comment (jnyrup) [inline Src/FluentAssertions/Equivalency/EquivalencyValidator.cs]: `shouldRecurse` is always `true` at this point, so the function could return `void` instead.

File: Src/FluentAssertions/Equivalency/EquivalencyValidator.cs [hunk 2]
```diff
@@ -44,87 +43,89 @@ public void AssertEquality(EquivalencyValidationContext context)
...
 
-        private static bool HasReachedMaximumRecursionDepth(string propertyPath)
+        private void RunStepsUntilEquivalencyIsProven(IEquivalencyValidationContext context)
         {
-            int depth = propertyPath.Count(chr => chr == '.');
+            foreach (var step in AssertionOptions.EquivalencySteps)
+            {
+                if (step.CanHandle(context, config) && step.Handle(context, this, config))
+                {
+                    return;
```
Comment (jnyrup) [inline Src/FluentAssertions/Equivalency/EquivalencyValidator.cs]: 👍 I've always found the `wasHandled = true` to be smelly.

File: Src/FluentAssertions/Equivalency/EquivalencyValidator.cs [hunk 3]
```diff
@@ -44,87 +43,89 @@ public void AssertEquality(EquivalencyValidationContext context)
...
+                    RunStepsUntilEquivalencyIsProven(context);
+                }
+            }
+        }
+
+        private bool ShouldCompareMembersThisDeep(string selectedMemberPath)
+        {
+            const char memberSeparator = '.';
+            var depth = selectedMemberPath.Count(chr => chr == memberSeparator);
+            bool shouldRecurse = config.AllowInfiniteRecursion || depth < MaxDepth;
```
[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Src/FluentAssertions/Equivalency/EquivalencyValidator.cs]: I experimented with various refactorings and liked this the most.

File: Src/FluentAssertions/Equivalency/EquivalencyValidator.cs [hunk 4]
```diff
@@ -44,87 +43,89 @@ public void AssertEquality(EquivalencyValidationContext context)
...
+            foreach (var step in AssertionOptions.EquivalencySteps)
+            {
+                if (step.CanHandle(context, config) && step.Handle(context, this, config))
+                {
+                    return;
+                }
+            }
 
-            return (depth >= MaxDepth);
+            throw new NotImplementedException($"No {nameof(IEquivalencyStep)} was found to handle the context. ");
```
[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Src/FluentAssertions/Equivalency/EquivalencyValidator.cs]: No, this should never happen. If it does, somebody has removed some of the built-in support for certain types.

File: Src/FluentAssertions/Equivalency/EquivalencyValidator.cs (hunk 1 above)
Comment (dennisdoomen) [reply Src/FluentAssertions/Equivalency/EquivalencyValidator.cs]: Unless somebody wrapped the call to `BeEquivalent` in another `AssertionScope`

File: Src/FluentAssertions/Execution/ContextDataItems.cs [hunk 5]
```diff
@@ -65,25 +60,26 @@ public T Get<T>(string key)
...
-            private readonly Reportability reportability;
-
-            public DataItem(string key, object value, Reportability reportability)
+            public DataItem(string key, object value, bool reportable, bool requiresFormatting)
             {
                 Key = key;
                 Value = value;
-                this.reportability = reportability;
+                Reportable = reportable;
+                RequiresFormatting  = requiresFormatting;
```
Comment (jnyrup) [inline Src/FluentAssertions/Execution/ContextDataItems.cs]: 🙃 double whitespace

File: Src/FluentAssertions/Execution/ContextDataItems.cs [hunk 6]
```diff
@@ -65,25 +60,26 @@ public T Get<T>(string key)
...
-            public string Key { get; private set; }
+            public string Key { get; }
+
+            public object Value { get; }
 
-            public object Value { get; private set; }
+            public bool Reportable { get; }
 
-            public bool Reportable => reportability == Reportability.Reportable;
+            public bool RequiresFormatting {get; }
```
Comment (jnyrup) [inline Src/FluentAssertions/Execution/ContextDataItems.cs]: 🙃 missing whitespace `{get`

File: Src/FluentAssertions/Equivalency/EquivalencyValidator.cs [hunk 7]
```diff
@@ -44,87 +43,86 @@ public void AssertEquality(EquivalencyValidationContext context)
...
+            bool isComplexType = IsComplexType(context.Expectation);
+
+            return objectTracker.IsCyclicReference(new ObjectReference(context.Expectation, context.SelectedMemberPath, isComplexType));
         }
 
-        private bool ContinueRecursion(string memberAccessPath)
+        private bool IsComplexType(object expectation)
         {
-            if (config.AllowInfiniteRecursion || !HasReachedMaximumRecursionDepth(memberAccessPath))
+            if (expectation == null)
```
Comment (jnyrup) [inline Src/FluentAssertions/Equivalency/EquivalencyValidator.cs]: 🙃 `is null`

File: Src/FluentAssertions/Execution/AssertionScope.cs [hunk 8]
```diff
@@ -1,16 +1,17 @@
-﻿#region
-
+﻿
```
Comment (jnyrup) [inline Src/FluentAssertions/Execution/AssertionScope.cs]: 🙃 empty line

File: Src/FluentAssertions/Equivalency/IEquivalencyValidator.cs [hunk 9]
```diff
@@ -2,6 +2,6 @@ namespace FluentAssertions.Equivalency
 {
     public interface IEquivalencyValidator
     {
-        void AssertEqualityUsing(IEquivalencyValidationContext context);
+        void RecursivelyAssertEquality(IEquivalencyValidationContext context);
```
Comment (jnyrup) [inline Src/FluentAssertions/Equivalency/IEquivalencyValidator.cs]: @dennisdoomen renaming this is a breaking change and should be reverted.

Comment (dennisdoomen) [reply Src/FluentAssertions/Equivalency/IEquivalencyValidator.cs]: How so?

Comment (jnyrup) [reply Src/FluentAssertions/Equivalency/IEquivalencyValidator.cs]: From the CoreFX [Breaking Change Rules]((https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md#members))
>Removing or renaming a member, including a getter or setter from a property or enum members

Though I can't find any code on the internet that implements `IEquivalencyValidator` except `EquivalencyValidator`, [Every change breaks someone's workflow](https://xkcd.com/1172/) and last time we thought it didn't matter, it [turned out bad](https://github.com/fluentassertions/fluentassertions.json/issues/27).
I think the benefit of renaming a public method versus the risk of introducing a breaking change, does not seem justified.

**Pull request merged**